### PR TITLE
Improvement of setup.py and license!

### PR DESCRIPTION
**Pull request by otizonaizit**
In this pull request I am improving a bit the `setup.py` to have all the metadata in a single place, also visible after the package has been installed.
I also chose a GPLv3+ license. If you don't agree please leave a comment on this PR ;)

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -1,115 +1,54 @@
...
+                    if node.targets[0].id == key:
+                        return node.value.s
+                except:
+                    pass
+    # return "not available" if we didn't find the variable
+    return 'N/A'
+
+# pick the relevant keywords from the __init__.py file
+metadata_vars = ('name', 'version', 'description', 'author', 'license', 'url')
+metadata = dict((var, parse_keyword('__%s__'%var)) for var in metadata_vars)
```
Comment (pberkes) [inline setup.py:27]: This is to avoid `from grader import __version__, __author__, ...` and `setup(version=__version__, author=__author__, ...)` right? 😉

**Comment by pberkes** [comment]
LGTM, including license choice

**Comment by otizonaizit** [comment]
>+# pick the relevant keywords from the __init__.py file
>+metadata_vars = ('name', 'version', 'description', 'author', 'license', 'url')
>+metadata = dict((var, parse_keyword('__%s__'%var)) for var in metadata_vars)
>
>This is to avoid `from grader import __version__, __author__, ...` and `setup(version=__version__, author=__author__, ...)` right? 😉
yes, and frankly I don't understand why there's no function in setuptools to do this automatically. Importing on installation I assume is a no-go (how do you know about side-effects of importing?). The only other option is to duplicate the metadata in the setup.py file, if you want to have the metadata available in the installed library too…

**Comment by pberkes** [comment]
I think it's pretty standard to import that information from the main package in `setup.py`. The `__init__.py` in those cases is empty except for that information, so it's safe.

However, it's not recommended to execute `git` in `__init__.py`... how about getting the revision number in `setup.py`? Or wrapping that code in a function so if necessary you can call it?

**Comment by keszybz** [comment]
LGTM.

**Comment by otizonaizit** [comment]
>I think it's pretty standard to import that information from the main package in `setup.py`. The `__init__.py` in those cases is empty except for that information, so it's safe.
well, in our case it may be safe, but how do you know in general?

>However, it's not recommended to execute `git` in `__init__.py`... how about getting the revision number in `setup.py`? Or wrapping that code in a function so if necessary you can call it?
well, numpy somehow manages to have a __git_revision__ module attribute on a standard installation, but maybe they read it from a module that is automatically generated during the setup.py call. If you find a way to do that without messing up with setuptools internals I would be happy ;)

**Comment by pberkes** [comment]
In general the people who write `setup.py` are the same as those who write `__init__.py`.

In numpy they have a function in `setup.py`: https://github.com/numpy/numpy/blob/master/setup.py#L108

Really, executing `git` every time that you import the package or run the tests is not a good idea.

**Comment by jakobj** [comment]
I always got the version number like this

```python
with open(INITFILE, 'r') as f:
    for l in f:
        if '__version__' in l:
            try:
                version = re.compile('[0-9]+.[0-9]+.[0-9]+').search(l).group()
            except AttributeError:
                raise ValueError('Could not determine package version.')
            else:
                break
```

and now I feel stupid ;)

Agreeing with the licence 👍

**Comment by keszybz** [comment]
This seems overkill too. `re` caches expressions, so `re.search()` would be just as effective as `re.compile.().search()`.  And also being too strict isn't useful here. If somebody set's the version to 0.1.0rc1, the code should work too.

```python
with open(INITFILE, 'rt') as f:
    for line in f:
         m = re.match(r'^__version__\s*=\s*(\S+)', line)
         if m:
                  return m.group(1)
    else:
        raise ValueError('__version__ not found')
```
?

**Comment by otizonaizit** [comment]
>This seems overkill too. `re` caches expressions, so `re.search()` would be just as effective as `re.compile.().search()`.  And also being too strict isn't useful here. If somebody set's the version to 0.1.0rc1, the code should work too.
but it's not only about the version, is also about a lot of other metadata… anyone willing to have a go at just reading the __init__ file and parse the relevant metadata? just post it as an update to this PR

**Comment by otizonaizit** [comment]
>In general the people who write `setup.py` are the same as those who write `__init__.py`.
exactly, so they forget to update either setup.py or __init__.py with the right version number or copyright year… maybe it's me, but I really can not stand duplication of metadata, it hurts my inner symmetry needs. As you have seen, I am willing to go through the AST to satisfy my inner needs :P

>In numpy they have a function in `setup.py`: https://github.com/numpy/numpy/blob/master/setup.py#L108
>
>Really, executing `git` every time that you import the package or run the tests is not a good idea.
OK, agreed. Can you update the PR with a function similar to the numpy one (although we need a much less fancy one)? Given that we only support Py3 there's no need to use fancy decoding, nor to parse the version number in any particular way…

**Comment by pberkes** [comment]
>> In general the people who write `setup.py` are the same as those who write `__init__.py`.

>exactly, so they forget to update either setup.py or __init__.py with the right version number or copyright year… maybe it's me, but I really can not stand duplication of metadata, it hurts my inner symmetry needs. As you have seen, I am willing to go through the AST to satisfy my inner needs :P

What I meant is that since you control both, you can safely `from grader import __version__` in `setup.py`

**Comment by pberkes** [comment]
> OK, agreed. Can you update the PR with a function similar to the numpy one (although we need a much less fancy one)? Given that we only support Py3 there's no need to use fancy decoding, nor to parse the version number in any particular way…

Not before Thursday I'm afraid

**Comment by keszybz** [comment]
> What I meant is that since you control both, you can safely from grader import __version__ in setup.py

That'd be my preference too.

File: setup.py (hunk 1 above)
Comment (jni) [reply setup.py:27]: A dict comprehension is more readable and concise imho:

```python
metadata = {var: parse_keyword('__%s__' % var) for var in metadata_vars}
```

or, if you want to support only 3.6+:

```python
metadata = {var: parse_keyword(f'__{var}__') for var in metadata_vars}
```

**Comment by jni** [comment]
I don't consider this "my" project so I'm happy to defer to the core contributors regarding the licensing. I would however like to repeat my current spiel about licenses: when in Rome, do as the Romans do. The scientific Python ecosystem is almost universally BSD, and won't touch GPL. See the latest spat about this:

https://mail.python.org/pipermail/scikit-image/2017-November/005407.html

So, regardless of your thoughts about copyleft vs permissive, in my view, community trumps all other considerations. The community, after all, is what makes SciPy so attractive. So, my vote is for BSD (I left MIT there because they are much of a muchness and it was the default in the PyPA template), but I will not stand in the way of GPL if you all still prefer that choice.

Regarding version numbers, I like the current solution, but I thought I would make you all aware of this library in case you haven't encountered it yet: https://github.com/peritus/bumpversion

I've used it and it's great.

**Comment by pberkes** [comment]
@jni for a library I would  go for BSD, but this is an end-user tool and I feel like we can be copyleftist in this instance

`bumpversion` is interesting, this could be a good opportunity to try it out! I volunteer giving it a go tomorrow, unless @otizonaizit objects

**Comment by otizonaizit** [comment]
On 8 November 2017 08:50:01 CET, Pietro Berkes <notifications@github.com> wrote:
>@jni for a library I would  go for BSD, but this is an end-user tool
>and I feel like we can be copyleftist in this instance
I agree!

>`bumpversion` is interesting, this could be a good opportunity to try
>it out! I volunteer giving it a go tomorrow, unless @otizonaizit
>objects
Yes, but then how does it work when you run setup.py when the tool is not installed? (Same objection for importing grader in setup.py: what happens if we have dependencies and they are not installed when you first run setup.py: AFAIK pip can install dependencies while running the setup function from setup.py, not before...)

**Comment by pberkes** [comment]
If I understand correctly, bump version is an external tool, not a
dependency.

As for importing from grader, the package is in the current directory so
that is not an issue in my experience. __init__ normally does not import
dependencies

**Comment by otizonaizit** [comment]
i had a look at `bumpversion`, and it seems to me that it is frankly overkill for grader. Come on, you can't claim it's better than my 5 lines of AST analysis for a little script :P

**Comment by pberkes** [comment]
I'll merge as it is (it is functional), and give my proposal in a separate PR

**Pull request merged**